--- util/grub-install.in.orig	2002-05-20 13:21:50.000000000 +0200
+++ util/grub-install.in	2003-09-05 21:20:35.000000000 +0200
@@ -27,7 +27,7 @@
 host_cpu=@host_cpu@
 host_os=@host_os@
 host_vendor=@host_vendor@
-pkgdatadir=${datadir}/${PACKAGE}/${host_cpu}-${host_vendor}
+pkgdatadir=${datadir}/${PACKAGE}
 
 grub_shell=${sbindir}/grub
 log_file=/tmp/grub-install.log.$$
@@ -378,15 +378,6 @@
     exit 1
 fi
 
-# Copy the GRUB images to the GRUB directory.
-for file in ${grubdir}/stage1 ${grubdir}/stage2 ${grubdir}/*stage1_5; do
-    rm -f $file || exit 1
-done
-for file in \
-    ${pkgdatadir}/stage1 ${pkgdatadir}/stage2 ${pkgdatadir}/*stage1_5; do
-    cp -f $file ${grubdir} || exit 1
-done
-
 # Make sure that GRUB reads the same images as the host OS.
 test -n "$mkimg" && img_file=`$mkimg`
 test -n "$mklog" && log_file=`$mklog`
