--- ../util/grub-set-default.in~	2004-07-16 14:44:57.000000000 +0300
+++ ../util/grub-set-default.in	2005-07-12 20:57:17.000000000 +0300
@@ -59,7 +59,7 @@
 	;;
     *)
 	if test "x$entry" != x; then
-	    echo "More than one entries?" 1>&2
+	    echo "More than one entry?" 1>&2
 	    usage
 	    exit 1
 	fi
@@ -90,7 +90,6 @@
 
 file=${grubdir}/default
 if test -f ${file}; then
-    chmod 0600 ${file}
     rm -f ${file}
 fi
 cat <<EOF > $file
@@ -106,9 +105,10 @@
 #
 #
 # WARNING: If you want to edit this file directly, do not remove any line
-# from this file, including this warning. Using \`grub-set-default\' is
+# from this file, including this warning. Using \`grub-set-default' is
 # strongly recommended.
 EOF
+chmod 0600 ${file}
 
 # Bye.
 exit 0
--- ../util/grub-set-default.in~	2005-07-12 20:58:25.000000000 +0300
+++ ../util/grub-set-default.in	2005-07-12 21:48:59.000000000 +0300
@@ -35,7 +35,7 @@
   -v, --version           print the version information and exit
   --root-directory=DIR    Use the directory DIR instead of the root directory
 
-ENTRY is a number or the special keyword \`default\'.
+ENTRY is a number or the special keyword \`default'.
 
 Report bugs to <bug-grub@gnu.org>.
 EOF
