--- grub-0.97/netboot/natsemi.c.orig	2003-07-09 13:45:38.000000000 +0200
+++ grub-0.97/netboot/natsemi.c	2005-06-12 15:55:59.000000000 +0200
@@ -608,7 +608,7 @@
 		 const char  *p)     /* Packet */
 {
     u32 to, nstype;
-    u32 tx_status;
+    volatile u32 tx_status;
     
     /* Stop the transmitter */
     outl(TxOff, ioaddr + ChipCmd);
@@ -647,7 +647,7 @@
 
     to = currticks() + TX_TIMEOUT;
 
-    while ((((volatile u32) tx_status=txd.cmdsts) & OWN) && (currticks() < to))
+    while (((tx_status = txd.cmdsts) & OWN) && (currticks() < to))
         /* wait */ ;
 
     if (currticks() >= to) {
--- grub-0.97/netboot/sis900.c~	2009-01-12 13:41:30.000000000 +0100
+++ grub-0.97/netboot/sis900.c	2009-01-12 13:42:51.000000000 +0100
@@ -1081,7 +1081,7 @@
                 const char  *p)     /* Packet */
 {
     u32 to, nstype;
-    u32 tx_status;
+    volatile u32 tx_status;
     
     /* Stop the transmitter */
     outl(TxDIS | inl(ioaddr + cr), ioaddr + cr);
@@ -1120,7 +1120,7 @@
 
     to = currticks() + TX_TIMEOUT;
 
-    while ((((volatile u32) tx_status=txd.cmdsts) & OWN) && (currticks() < to))
+    while (((tx_status = txd.cmdsts) & OWN) && (currticks() < to))
         /* wait */ ;
 
     if (currticks() >= to) {
