Index: stage2/builtins.c
===================================================================
RCS file: /cvsroot/grub/grub/stage2/builtins.c,v
retrieving revision 1.142
diff -u -r1.142 builtins.c
--- stage2/builtins.c	10 Dec 2002 17:30:29 -0000	1.142
+++ stage2/builtins.c	4 Mar 2003 19:54:41 -0000
@@ -4188,7 +4188,7 @@
 	      
 	      if (! grub_memcmp (arg, name, len))
 		{
-		  grub_strcpy (options[i].var, arg + len);
+		  grub_strcpy (options[i].var, ti_unescape_string(arg + len));
 		  break;
 		}
 	    }
@@ -4215,11 +4215,11 @@
       /* No option specifies printing out current settings.  */
       term = ti_get_term ();
 
-      grub_printf ("name=%s\n", term.name);
-      grub_printf ("cursor_address=%s\n", term.cursor_address);
-      grub_printf ("clear_screen=%s\n", term.clear_screen);
-      grub_printf ("enter_standout_mode=%s\n", term.enter_standout_mode);
-      grub_printf ("exit_standout_mode=%s\n", term.exit_standout_mode);
+      grub_printf ("name=%s\n", ti_escape_string(term.name));
+      grub_printf ("cursor_address=%s\n", ti_escape_string(term.cursor_address));
+      grub_printf ("clear_screen=%s\n", ti_escape_string(term.clear_screen));
+      grub_printf ("enter_standout_mode=%s\n", ti_escape_string(term.enter_standout_mode));
+      grub_printf ("exit_standout_mode=%s\n", ti_escape_string(term.exit_standout_mode));
     }
 
   return 0;
Index: stage2/terminfo.h
===================================================================
RCS file: /cvsroot/grub/grub/stage2/terminfo.h,v
retrieving revision 1.1
diff -u -r1.1 terminfo.h
--- stage2/terminfo.h	29 Nov 2002 20:39:24 -0000	1.1
+++ stage2/terminfo.h	4 Mar 2003 19:54:41 -0000
@@ -21,7 +21,7 @@
 #ifndef GRUB_TERMCAP_HEADER
 #define GRUB_TERMCAP_HEADER	1
 
-#define TERMINFO_LEN 32
+#define TERMINFO_LEN 40
 
 typedef struct terminfo
 {
