diff -ruN grub-0.96./util/grub-install.in grub-0.96/util/grub-install.in
--- grub-0.96./util/grub-install.in	2005-02-01 22:18:17.000000000 +0100
+++ grub-0.96/util/grub-install.in	2005-02-01 22:18:52.579617408 +0100
@@ -27,7 +27,7 @@
 host_cpu=@host_cpu@
 host_os=@host_os@
 host_vendor=@host_vendor@
-pkglibdir=${libdir}/${PACKAGE}/${host_cpu}-${host_vendor}
+pkglibdir=${libdir}/${PACKAGE}
 
 grub_shell=${sbindir}/grub
 grub_set_default=${sbindir}/grub-set-default
@@ -406,15 +406,6 @@
     exit 1
 fi
 
-# Copy the GRUB images to the GRUB directory.
-for file in ${grubdir}/stage1 ${grubdir}/stage2 ${grubdir}/*stage1_5; do
-    rm -f $file || exit 1
-done
-for file in \
-    ${pkglibdir}/stage1 ${pkglibdir}/stage2 ${pkglibdir}/*stage1_5; do
-    cp -f $file ${grubdir} || exit 1
-done
-
 # Make a default file.
 ${grub_set_default} --root-directory=${rootdir} default
 
